enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,701 @@
+OVN-Kubernetes CRD:
+
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: RouteAdvertisements
+metadata:
+  name: default
+spec:
+    # networkSelector:
+    # nodeSelector:
+    # frrConfigurationSelector:
+    targetVRF: default
+    advertisements:
+      podNetwork: true
+      egressIP: true
+```      
+
+In the above example, an optional networkSelector may also be optionally supplied which will match namespaces/networks.
+When omitted, RouteAdvertisements will be applied to the default cluster network. If a selector is used that selects the
+default cluster network, it will be enabled for any namespace using the cluster default network. Additionally,
+advertisements may be limited to specific nodes via the nodeSelector.
+
+A networkSelector may select more than one network, including user-defined networks. In such a case the network subnets
+will be checked by OVN-Kubernetes to determine if there is any overlap of the IP subnets. If so, an error status will be
+reported to the CRD and no BGP configuration will be done by OVN-Kubernetes.
+
+The CRD will support enabling advertisements for pod subnet and egress IPs. Note, MetalLB still handles advertising
+LoadBalancer IP so there is no conflict of responsibilities here. When the pod network is set to be advertised, there is
+no longer a need to SNAT pod IPs for this network to the node IP. Therefore, when pod network advertisements are enabled,
+the traffic from these pods will no longer be SNAT'ed on egress.
+
+The "targetVRF" key is used to determine which VRF the routes should be advertised in. The default value is "auto", in
+which case OVN-Kubernetes will advertise routes in the VRF that corresponds to the selected network. Alternatively, the
+user may specify the name of the VRF, which would cause routes to be leaked for this network onto that VRF. One use case
+for this would be when a user wants to define a network with a specific IP addressing scheme, and then wants to advertise
+the pod IPs into the provider BGP network without VPN. By specifying the targetVRF as "default", routes will be leaked
+into the default VRF. Note that by using route leaking with a user-defined network, the network is no longer fully
+isolated, as now any other networks also leaked or attached to that VRF may reach this user-defined network. If a user
+attempts to leak routes into a targetVRF for a user-defined network whose IP subnet would collide with another,
+OVN-Kubernetes will report an error to the RouteAdvertisement status.
+
+The frrConfigurationSelector is used in order to determine which FRRConfiguration CR to use for building the OVN-Kubernetes
+driven FRRConfiguration. OVN-Kubernetes needs to leverage a pre-existing FRRConfiguration to be able to find required
+pieces of configuration like BGP peering, etc. If more than one FRRConfiguration is found matching the selector, then
+an error will be propagated to the RouteAdvertisements CR and no configuration shall be done.
```

Comment:
If this is not actually `frr`-specific configuration then it should have a more generic name. If it _is_ `frr`-specific then it probably shouldn't be. We should be providing high-level configuration, not escape hatches for people to directly configure the low-level implementation (which are likely to lead to compatibility problems and customers doing things we didn't intend for them to be doing).

Reply:
The two options here were either duplicating what we have in the frrconfiguration crd (mostly related to how to establish a bgp session with a given neighbor, such as ip, password, asn, timers and eventually bfd) or to rely to the existing configuration and select it, to avoid having multiple ways to describe the same thing.

On the other hand an ovnk wrapper will make the api it more sealed and will open the door to easily change the underlying bgp implementation.

Reply:
The FRR technology is part of the API (FRR-K8S), and OVNK is generating FRR-K8S API from other FRR-K8S CRs, hence whey the selector is there. I guess if we want to abstract from FRR in OVNK, we would need a more generic construct like:

```
RouteAdvertisements:
  sourceRoutingConfiguration:
    type: frr-k8s
    selector:
          select: something
```

is this what you are thinking @danwinship ?

Reply:
I mean
1. If we aren't 100% certain that we will always want to be using frr as the backend, then we shouldn't be exposing the fact that we currently are
2. If we don't explicitly intend for the admin to be able to use every possible frr feature (including future ones we don't know about yet), then we shouldn't expose every possible frr config option.

Like, the `network.operator.openshift.io` object lets you configure arbitrary kube-proxy options and that was always somewhat problematic for openshift-sdn, because the admin might change the proxy behavior in some way we weren't really expecting or accounting for. Or with `ingress-nginx`, apparently some annotation values get substituted directly into the generated nginx config, and this has been a big source of CVEs, because there's no way ingress-nginx can understand every possible nginx config option to make sure you're using it in a way that makes sense from ingress-nginx.

Reply:
I see your point, but that is kind of outside the scope of this enhancement. FRR-K8S is a feature that is going to be GA'ed in 4.17 with or without OVNK using it. From the OVNK side we are restricting what config we will allow configured in FRR by using the routeAdvertisements CRD. However, nothing stops a user from going and creating an FRR-K8S rawConfig type CR, and putting whatever config they want in it.

@fedepaol have you thought about what support will look like for FRR-K8S in the case where a user does rawConfig?

Reply:
> However, nothing stops a user from going and creating an FRR-K8S rawConfig type CR, and putting whatever config they want in it.

I assume you mean that _cluster admins_ can do that, not anyone else, right?

If frr-k8s lets cluster admins specify arbitrary config, then that seems bad for support reasons, but it's not a security problem (and it's out of scope for this enhancement I guess).

But we need to make sure that in the case where a namespace admin creates a user-defined primary network for their namespace, that they can't also configure routing in any ways that the cluster admin didn't intend for them to. (Maybe this is already covered... I didn't check all the details of what config is allowed in what cases...)

Reply:
The frr-k8s is very limited in what it allows to do (basically today it allows to do what metallb does, advertising prefixes, plus the ability of receiving them too). 

The rawConfig field is not being documented and there's no plan to support it, but it's there to allow experimenting and as a backdoor in case there are scenarios to be solved quickly. If a user wants to use rawConfig, there'll be a support excpetion and a clear agreement on what they can add and why they can use it. 

Finally, the frr-k8s daemon api is namespace scoped, and the instance ignores entries that do not belong to the namespace created by CNO for it, so it's definitely only for the cluster admin.